### PR TITLE
Set target dir, copy file url, more file info, other changes

### PR DESCRIPTION
**Pull request by olfek**
### This PR introduces the following features

1. **Ability to set target directory (home/root/base path)**
2. More file info in the file info popup, which is useful for smaller devices such as mobile phones which cannot display the extra columns.
3. Configurable hashing algorithms.
Some people may only want to allow md5 for example to reduce load on their servers, or with this option, be able to confidentially increase the `MAX_HASH_SIZE` by choosing less compute intensive algorithms such as md5.
4. The page title no longer contains `•`.
This means people can use whatever symbol they want and in the case of an empty `SITE_TITLE`, their is no trailing symbol.
5. Ability to copy the URL of a file.
This is useful for when a media file is in a format that the browser does not support, `.mkv` for example, the URL can be copy and pasted into a media streaming app such as VLC.

Other changes include minor fixes and related test fixes.

### Example NGINX server configuration

```
server {
    ...

    # Location of PHP Directory Lister
    set $my_root /www;

    listen 80;

    # PHP Directory Lister
    location / { 
        root $my_root;
        index /index.php;
    }

    # Serve static files
    location /files/ {
        rewrite ^/files/(.*)$ /$1 break;
        root <path_to_target_directory>; 
    }

    # Process PHP files
    location ~* \.php$ {
        fastcgi_pass unix:/run/php-fpm/www.sock;
        fastcgi_param   SCRIPT_FILENAME    $my_root$fastcgi_script_name;
        fastcgi_param   SCRIPT_NAME        $fastcgi_script_name;
        include         fastcgi_params;
    }

    ...
}
```
### Example `.env` file
```
...

FILE_URL_PREFIX=/files
BASE_PATH=<path_to_target_directory>

...
```

### CODE REVIEW

**Review by PHLAK — CHANGES_REQUESTED** [review]
Please break these changes up into individual pull requests (one PR per feature).

**Comment by olfek** [comment]
Hello, I only have one squashed commit, to create multiple PRs I will need to break it down into smaller commits first.

I'm concerned that doing this may break things and it's worth noting that the current state of the code has been throughly tested manually.

I will update the description of the PR this evening so you have a better idea of the work that has been done.

Thank you for your prompt response 🙂

**Comment by olfek** [comment]
@PHLAK I have added a description :slightly_smiling_face:

**Comment by PHLAK** [comment]
First, thank you for the PR.  I appreciate the time and effort that goes into every PR. Unfortunately, I will not be accepting this combined pull request. If you wish to separate this into separate PRs I will review them individually. I will, however give my initial thoughts about each feature without digging in too deep.

### 1. Ability to set target directory (home/root/base path)

I'd love to find a way to do this but would need to review the individual PR for this feature before making any decisions.

### 2. More file info in the file info popup

What additional info?

### 3. Configurable hashing algorithms.

This isn't something I wish to change at this time.

### 4. The page title no longer contains •.

>  in the case of an empty SITE_TITLE, their is no trailing symbol

While I agree it would be nice to hide the bullet if `SITE_TITLE` is empty the bullet is used as a separator between the path and site title when present and I don't want to remove it from the code.

### 5. Ability to copy the URL of a file.

Would need to review the individual PR for this feature before making any decisions.

**Comment by olfek** [comment]
Hello, why has this been closed? @PHLAK

File: app/src/Controllers/FileInfoController.php [hunk 1]
```diff
@@ -38,21 +39,21 @@ public function __invoke(Request $request, Response $response): ResponseInterfac
...
-            return $response->withStatus(500, $this->translator->trans('error.file_size_exceeded'));
-        }
-
         $response->getBody()->write($this->cache->get(
             sprintf('file-info-%s', sha1($file->getRealPath())),
             function () use ($file): string {
-                return json_encode(['hashes' => $this->calculateHashes($file)]);
+                return json_encode(
+                    ['modification time' => date($this->config->get('date_format'), $file->getMTime())] +
+                    ['size' => Utils::sizeForHumans($file)] +
```
Comment (olfek) [inline app/src/Controllers/FileInfoController.php:54]: @PHLAK Regarding feature number 2
> What additional info?

This additional info

**Comment by olfek** [comment]
@PHLAK Why do you not wish to have feature number 3?

**Comment by olfek** [comment]
@PHLAK Regarding feature number 4
> While I agree it would be nice to hide the bullet if SITE_TITLE is empty the bullet is used as a separator between the path and site title when present and I don't want to remove it from the code.

Now that the bullet has been removed, users can add their own symbol into `SITE_TITLE` if they want to have a separator, this information can be added to the upgrade notes.

File: app/config/app.php [hunk 2]
```diff
@@ -27,7 +35,7 @@
      *
      * Default value: 'Directory Lister'
      */
-    'site_title' => DI\env('SITE_TITLE', 'Directory Lister'),
+    'site_title' => DI\env('SITE_TITLE', '• Directory Lister'),
```
Comment (olfek) [inline app/config/app.php:38]: The bullet has been moved here, so out of the box installations will still have a separator @PHLAK

**Comment by olfek** [comment]
@PHLAK Also, FYI, when developing, and targeting an external directory, you want to set `FILE_URL_PREFIX` to a `localhost` address with a port different from the one the app is running on, then, start another server at that port for the sole purpose of serving files.

**Comment by olfek** [comment]
@PHLAK I'd greatly appreciate if you could re-open this PR as I have spent over a week working on it 🙂

**Comment by PHLAK** [comment]
@olfek I appreciate the time and effort that went into this and would like to work on getting some of this code merged into Directory Lister.

However, as I stated before I will not be accepting this combined PR. If you would separate this into individual PRs (one per feature) I can review and discuss them individually.

**Comment by olfek** [comment]
> Hello, I only have one squashed commit, to create multiple PRs I will need to break it down into smaller commits first.
> 
> **I'm concerned that doing this may break things and it's worth noting that the current state of the code has been thoroughly tested manually.**
> 
> I will update the description of the PR this evening so you have a better idea of the work that has been done.
> 
> Thank you for your prompt response 🙂

@PHLAK See my earlier comment about this above, and besides, I don't want to spend anymore time on this as I have other things to do, all of the features I have offered to you in this PR _should_ be accepted in my opinion, they are all great improvements.

If you spend some time looking at the code, you will begin to feel more comfortable with the changes I have made 🙂

**Comment by PHLAK** [comment]
I have looked at the code and wish to discuss some changes. However, I will not be doing that unless this PR is broken into individual PRs per feature. I will not ask again.

**Comment by olfek** [comment]
@PHLAK Why can't you just review everything in this PR?

**Comment by olfek** [comment]
@PHLAK You trying to make me do extra "red tape" work, don't be selective about the features and just review everything please 🙂

**Comment by olfek** [comment]
Lets start to discuss the actual code @PHLAK 🙂 I look forward to seeing your comments on the code

**Comment by olfek** [comment]
Hello? @PHLAK

**Comment by olfek** [comment]
I really don't want to have to maintain a fork of this @PHLAK

**Comment by PHLAK** [comment]
@olfek I have stated multiple times that I will review each feature individually once broken into separate PRs. This requirement is clearly defined in our [contributing guidelines](https://github.com/DirectoryLister/DirectoryLister/blob/master/.github/CONTRIBUTING.md) and I will not be making an exception in this case. Please stop spamming.

**Pull request closed without merging**